### PR TITLE
Different Loss Ablations Algorithms

### PR DESCRIPTION
**Pull request by joshuakurien**
- FOND_N: contrastive loss with just non-overlapping class loss and no fairness with absolute error between NOC and OC
- FOND_NC: contrastive loss with just non-overlapping class loss, overall class loss and no fairness with absolute error between NOC and OC
- NOC: just non-overlapping class loss

### CODE REVIEW

File: domainbed/algorithms.py [hunk 1]
```diff
@@ -2361,6 +2346,194 @@ class FOND(XDomBase):
...
+        )
+        noc_class_loss = F.cross_entropy(
+            classifs, targets, weight=self.noc_weight.to(targets.device)
+        )
+        error_loss = oc_class_loss - noc_class_loss
+        if torch.isnan(error_loss):
+            error_loss = torch.tensor(0).to(targets.device)
+
+        class_loss = F.cross_entropy(classifs, targets)
+
```
Comment (kimathikaai) [inline domainbed/algorithms.py:2394]:
```python
if torch.isnan(noc_class_loss):
    noc_class_loss = torch.tensor(0).to(targets.device)
```

File: domainbed/algorithms.py [hunk 2]
```diff
@@ -2361,6 +2346,194 @@ class FOND(XDomBase):
...
```
Comment (kimathikaai) [inline domainbed/algorithms.py:2460]:
```python
if torch.isnan(noc_class_loss):
    noc_class_loss = torch.tensor(0).to(targets.device)
```

File: domainbed/algorithms.py [hunk 3]
```diff
@@ -2361,6 +2346,194 @@ class FOND(XDomBase):
...
```
Comment (kimathikaai) [inline domainbed/algorithms.py]:
```python
if torch.isnan(noc_class_loss):
    noc_class_loss = torch.tensor(0).to(targets.device)
```

File: domainbed/algorithms.py [hunk 4]
```diff
@@ -2361,6 +2346,194 @@ class FOND(XDomBase):
...
+        loss.backward()
+        self.optimizer.step()
+
+        return {
+            "loss": loss.item(),
+            "class_loss": class_loss.item(),
+            "xdom_loss": xdom_loss.item(),
+            "error_loss": error_loss.item(),
+            "mean_p": mean_positives_per_sample.item(),
+            "zero_p": num_zero_positives.item(),
```
Comment (kimathikaai) [inline domainbed/algorithms.py:2541]:
```python
"noc_class_loss": noc_class_loss.item(),
```

File: domainbed/algorithms.py [hunk 5]
```diff
@@ -2361,6 +2346,194 @@ class FOND(XDomBase):
...
```
Comment (kimathikaai) [inline domainbed/algorithms.py:2477]:
```python
"noc_class_loss": noc_class_loss.item(),
```

File: domainbed/algorithms.py [hunk 6]
```diff
@@ -2361,6 +2346,194 @@ class FOND(XDomBase):
...
```
Comment (kimathikaai) [inline domainbed/algorithms.py:2412]:
```python
"noc_class_loss": noc_class_loss.item(),
```

**Review by kimathikaai — CHANGES_REQUESTED** [review]

**Review by kimathikaai — APPROVED** [review]
